Track Shard-Snapshot Index Generation at Repository Root

original-brownbear opened the pull request:
### Changes to Root-Level index-N (RepositoryData)

This change adds a new field `"shards"` to `RepositoryData` that contains a mapping of `IndexId` to a `String[]`. This string array can be accessed by shard id to get the generation of a shard's shard folder (i.e. the `N` in the name of the currently valid `/indices/${indexId}/${shardId}/index-${N}` for the shard in question).

### Benefits

This allows for creating a new snapshot in the shard without doing any LIST operations on the shard's folder. In the case of AWS S3, this saves about 1/3 of the cost for updating an empty shard (see #45736) and removes one out of two remaining potential issues with eventually consistent blob stores (see #38941 ... now only the root `index-${N}` is determined by listing).

Also and equally if not more important, a number of possible failure modes on eventually consistent blob stores like AWS S3 are eliminated by moving all delete operations to the `master` node and moving from incremental naming of shard level index-N to uuid suffixes for these blobs.

### Only Master Deletes Blobs

This change moves the deleting of the previous shard level `index-${uuid}` blob to the master node instead of the data node allowing for a safe and consistent update of the shard's generation in the `RepositoryData` by first updating `RepositoryData` and then deleting the now unreferenced `index-${newUUID}` blob.
__No deletes are executed on the data nodes at all for any operation with this change.__

Note also: Previous issues with hanging data nodes interfering with master nodes are completely impossible, even on S3 (see next section for details).

### Why Move from index-${N} to index-${uuid} at the Shard Level

This change changes the naming of the shard level `index-${N}` blobs to a uuid suffix `index-${UUID}`. The reason for this is the fact that writing a new shard-level `index-` generation blob is not atomic anymore in its effect. Not only does the blob have to be written to have an effect, it must also be referenced by the root level `index-N` (`RepositoryData`) to become an effective part of the snapshot repository.
This leads to a problem if we were to use incrementing names like we did before. If a blob `index-${N+1}` is written but due to the node/network/cluster/... crashes the root level `RepositoryData` has not been updated then a future operation will determine the shard's generation to be `N` and try to write a new `index-${N+1}` to the already existing path. Updates like that are problematic on S3 for consistency reasons, but also create numerous issues when thinking about stuck data nodes.
Previously stuck data nodes that were tasked to write `index-${N+1}` but got stuck and tried to do so after some other node had already written `index-${N+1}` were prevented form doing so (except for on S3) by us not allowing overwrites for that blob and thus no corruption could occur.
Were we to continue using incrementing names, we could not do this. The stuck node scenario would either allow for overwriting the `N+1` generation or force us to continue using a `LIST` operation to figure out the next `N` (which would make this change pointless).
With uuid naming and moving all deletes to `master` this becomes a non-issue. Data nodes write updated shard generation `index-${uuid}` and `master` makes those `index-${uuid}` part of the `RepositoryData` that it deems correct and cleans up all those `index-` that are unused.

original-brownbear committed:
bck

original-brownbear committed:
sorta works

original-brownbear committed:
better

original-brownbear committed:
pass

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear committed:
moar asserts

original-brownbear committed:
fix CCR repo

original-brownbear committed:
fix tests

original-brownbear committed:
fix some inconsistencies

original-brownbear committed:
remove risky delete

original-brownbear committed:
nicer

original-brownbear committed:
extract fallback logic

original-brownbear committed:
comment

original-brownbear committed:
ensure not leaking shard index-N

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
some bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
half fix bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
assertion and nicer order

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
add version param

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
implement BwC

original-brownbear committed:
bulk delete old index-N

original-brownbear committed:
fix update on empty shard list

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
horrible but safe

original-brownbear committed:
drier

original-brownbear committed:
closer

original-brownbear committed:
stop redundant reading of metadata

original-brownbear committed:
nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
even safer ...

original-brownbear committed:
much faster

original-brownbear committed:
safer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
some doc improvements

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
start extracting index gen math

original-brownbear committed:
nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
simpler

original-brownbear committed:
smarter bwc

original-brownbear committed:
nicer

original-brownbear committed:
more sanity

original-brownbear committed:
fix more possible inconsistencies

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
fix tests

original-brownbear committed:
way moar docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
docs, better equals etc.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
update docs and add test todo

original-brownbear committed:
clarify

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
drier and nicer

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear committed:
slightly nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
nicer

original-brownbear committed:
nicer apis more docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
comment

original-brownbear committed:
nicer comment

original-brownbear committed:
nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
shorter diff

original-brownbear committed:
smaller diff

original-brownbear committed:
smaller diff

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Jenkins run elasticsearch-ci/2

original-brownbear committed:
nicer and fix stuff

original-brownbear committed:
smaller diff

original-brownbear committed:
smaller diff

original-brownbear committed:
nicer diff

original-brownbear committed:
smaller diff

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
yet smaller diff

original-brownbear committed:
save line

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
smaller diff

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
fix typo

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
fix

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Jenkins run elasticsearch-ci/2

original-brownbear commented:
Waiting on #47009 here

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
shorter

original-brownbear committed:
shorter

original-brownbear commented:
Jenkins run elasticsearch-ci/2 
Jenkins run elasticsearch-ci/bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Waiting on https://github.com/elastic/elasticsearch/pull/47293 here now which should bring the legacy and the new delete path in this PR much closer together code-wise :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
align new and old path

original-brownbear committed:
drier

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
shorter

original-brownbear committed:
simpler

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
simpller

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
https://github.com/elastic/elasticsearch/pull/47439 should shorten the diff here some more as well :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
shorter diff

original-brownbear committed:
shorter

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
more solid assertion

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
much shorter

original-brownbear committed:
shorter

original-brownbear committed:
shorter diff

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample (random Jenkins issue)

original-brownbear committed:
simpler

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
nicer

original-brownbear committed:
fix formatting

original-brownbear commented:
@ywelsch @tlrx I think I've now cleaned this up and shortened it as much as I could and it's ready for review. I know it's still a bit of a big one but I think it'll we be hard to split up the logic more without losing context.

In the end, it's just two parts here now:
1. The changed serialization for `RepositoryData` (that should be trivial)
2. The changed order of operations (did my best to update documentation here and share code with the BwC path so that we have minimal changes here)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
fix BwC

original-brownbear commented:
Hmm some SLM test failed here, looking into that. First time I'm seeing it.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
SLM failure is a bug in master, will open a PR for that

original-brownbear commented:
Jenkins run elasticsearch-ci/2

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -156,7 +157,7 @@ public void clusterChanged(ClusterChangedEvent event) {
                     cancelRemoved(currentSnapshots);
                     if (currentSnapshots != null) {
-                        startNewSnapshots(currentSnapshots);
+                        startNewSnapshots(currentSnapshots, event.state().nodes().getMinNodeVersion());
```
I think it's dangerous if the node decide whether to use old-style or new-style all by themselves as they sample the cluster state at different points in time, and therefore making different decisions. I would prefer for the master to establish this at the beginning of taking the snapshot, and then put "old-style" or "new-style" into the SnapshotInProgress.Entry of the cluster state, and all nodes just execute on that.

original-brownbear replied:
I figured it wouldn't matter since we never downgrade? Old-style is understood by all nodes and in case we only have new-style nodes in the CS at one point, that will never change will it?

But regardless, we can obviously track this in the CS if you want. Maybe we could just add a `version` field to `SnapshotInProgress.Entry`, then we can use it for follow-up changes BwC as well?

original-brownbear replied:
Update: I thought about this some more and I don't think it adds anything to track the style in the `SnapshotsInProgress.Entry` relative to the current solution.
The current solution can't move from a new version cluster back to a mixed version cluster which would be the only problematic case. In this case (or a failover from a new version to an old version master) the information about the style of snapshot would be lost anyway since the old version master can't track that new field. Obviously you would default it to old-style then, but it seems completely redundant and just adds more complication to `SnapshotsInProgress` serialization to no benefit?

ywelsch replied:
> The current solution can't move from a new version cluster back to a mixed version cluster

That's true. But for a particular snapshot, the transition can happen mid-way through the snapshot. That means that some data nodes write old-style, others new-style, and same goes for the master. I wonder if that will make things more complex than just sticking to one consistent style throughout a given snapshot.

original-brownbear replied:
> That means that some data nodes write old-style, others new-style, and same goes for the master. 

That's true, but note that creating a snapshot only has two steps on the repository level these days (we had three steps before since we wrote the clusterstate + index-metadata first but now we do that during finalization).
So even if data nodes write a mix, they will also send their new generation to `master` and `master` will finalize in the new format with some gens as numeric values instead of uuids.
The only maybe-issue I see is that we're currently not deleting the last numeric-suffix index-N on snapshot create (I could add that back, or we could ignore it since those get cleaned up during the next delete/cleanup anyway and don't take much space in the first place ... might not be worth the added code complexity).

original-brownbear replied:
Update:
The above was slighly stupid. I added back the BwC deletion of old index-N from primaries since obviously the mixed-cluster state could run for multiple snapshots and we'd continue to leak index-N in that case without that. I can't see any BwC case that isn't neatly handled now though.
Even if all data nodes were to write in old style and the master would finalize in new-style it'd be fine. Whether the master was in new or old-style mode when the whole snapshot started is entirely irrelevant to the snapshot process sine the master doesn't do any writes on snapshot initialization now.
Since it's impossible for a data-node to write a new-style index-${uuid} blob and the master to not finalize in the new-style I don't think any problems are possible here still.

original-brownbear replied:
> If put in SnapshotsInProgress, I would prefer a simple boolean. In master, that code can go away anyway.

`boolean` it is then :)

> I also wonder if by using null in ShardGenerations we can't distinguish between "there is no metadata file that we care about for this shard" and "the snapshot of this shard was taken while the cluster was on a version prior to 7.5.0".

This is a fair point. The current logic of trying to find a numeric `index-N` doesn't make sense in case of a newly created shared or index path. I'll put a placeholder for that case we can avoid the redundant listing of empty/new shard paths.

original-brownbear replied:
Added placeholder logic for known new shards in b9dbd7c5471 saving us a `LIST` for new indices (the removed BwC path in `RepositoryData` is irrelevant nowadays anyway).

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -323,6 +357,11 @@ public XContentBuilder snapshotsToXContent(final XContentBuilder builder) throws
+
+        if (version.onOrAfter(SnapshotsService.SHARD_GEN_IN_REPO_DATA_VERSION)) {
+            shardGenerations.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        }
```
`else` shardGenerations should be empty, right? Can we assert that?

original-brownbear replied:
Right, added that assertion.

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -384,47 +389,75 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            // Note: If we fail updating any of the individual shard paths, none of them are changed since the newly created
+            //       index-${gen_uuid} will not be referenced by the existing RepositoryData and new RepositoryData is only
+            //       written if all shard paths have been successfully updated.
+            deleteIndices(repositoryData, snapshotId, version, ActionListener.wrap(res -> {
```
the `deleteIndices` method name is confusing. That method does a lot of things.

original-brownbear replied:
Tried to name it better and added a comment on what this actually does, see if it's a little clearer now.

In fact, I'd like to just flatten the whole delete method into one (I had that in a previous version of this PR) to make the order of operations easier to follow. Just didn't want to do it here to keep the diff smaller, but I can if it helps ... seems to me it would make things nicer.

ywelsch replied:
The diff on these methods is already very complex (I checked out the code in my IDE to make sense of it), so perhaps just give it a go.

original-brownbear replied:
Flattened out the delete logic some more now and tried to add more comments and improve naming a little. Let me know if that helps.

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -595,51 +627,67 @@ private void deleteIndices(RepositoryData repositoryData, List<IndexId> indices,
+                        new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
+                    // Just invoke the listener without any shard generations to count it down, this index will be cleaned up
+                    // by the stale data cleanup in the end.
+                    deleteIdxMetaListener.onResponse(null);
```
What does this mean for newer ES versions where we rely on a new IndexShardSnapshot file to be written by this method?

ywelsch replied:
yes, I don't have a clear idea yet how this should be handled, but I now think that the current handling is safe. Let's keep as is (perhaps add a TODO).

ywelsch reviewed (changes requested):
Thanks for this work, Armin. I've done a first pass, but need to look more closely at some of this. I find the version-handling logic hard to follow (that it's doing the right thing). I've made a suggestion to simplify this for snapshots (determine old- or new-style mode at the beginning and stick with it (put it in SnapshotInProgress.Entry)). Something similar should be done for delete. Also I would like you to check if the code in the `deleteIndices` method can be restructured/simplified. It feels overly complex right now.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
CR comments

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -171,13 +192,20 @@ public RepositoryData withGenId(long newGeneration) {
+     *                                contains nodes from before {@link SnapshotsService#SHARD_GEN_IN_REPO_DATA_VERSION}.
      */
-    public RepositoryData removeSnapshot(final SnapshotId snapshotId) {
+    public RepositoryData removeSnapshot(final SnapshotId snapshotId, @Nullable final ShardGenerations updatedShardGenerations) {
```
Never, this was a leftover. Fixed now.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,168 @@
+     * @return indices for which shard generations are tracked
+     */
+    public List<IndexId> indices() {
+        return List.copyOf(shardGenerations.keySet());
```
Right ... sorry about that, cleaned that up now.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,168 @@
+            final Map<Integer, String> obsoleteShardIndices = new HashMap<>();
+            assert updatedGenerations != null
+                : "Index [" + indexId + "] present in previous shard generations, but missing from updated generations";
+            for (int i = 0; i < oldGens.size(); i++) {
```
> What if we have an index with 5 primaries, snapshot it, delete the index, and create the same index with 1 primary, and then snapshot it.

That's not possible. It would have a different index uuid and would get a different repo uuid. The shard count for indices in the repo is unchangable as far as I can tell.
The only situaiton where it could be increasing is if the first snapshot to an index failed for the highest shardId(s) but succeeded for some lower id(s). It's never decreasing though.

original-brownbear replied:
@ywelsch :man_facepalming: wow you're right I never noticed that (same name -> same indexUUID) for some reason. That's really unfortunate efficiency wise (not just in light of this change) ... though I guess it's an edge case.

I'll adjust he logic accordingly and add a test.

original-brownbear replied:
Adjusted code to handle this 0f4b316edfc and added test for the scenario (had to adjust the consistency checks to understand this as well)

original-brownbear replied:
@ywelsch `SnapshotsInProgress.Entry` now tracks the format of the snapshot 3a66c7966b3

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,168 @@
+                    final int size = shardIds.stream().mapToInt(i -> i).max().getAsInt() + 1;
+                    final String[] gens = new String[size];
+                    entry.getValue().forEach((shardId, generation) -> gens[shardId] = generation);
+                    return Arrays.asList(gens);
```
Yes, if a shard failed to update then `null` is used to signifie no change (or none exists yet, try to recover from numeric, if the previous generations had a `null` entry as well either due to being from a previous version or because the repo/index is new).

original-brownbear replied:
Sure done :)

original-brownbear committed:
Rename + assert

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -426,14 +466,38 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
+            final IndexId idx = indexLookup.get(indexId);
+            assert idx != null : "shard generations contained index [" + indexId + "] that isn't part of any snapshot";
+            for (int i = 0; i < gens.length; i++) {
+                builder.add(idx, i, gens[i]);
```
Yes in that case the shard-level meta is updated for those shards that did successfully snapshot so that we get the references to all the new data blobs and they can be used in the next snapshot.
This is basically the same thing we have now, except for the fact that a "rogue" primary that keeps writing to the shard directory but is otherwise disconnected from the master will simply write a `index-${uuid}` blob that master does not take into account and that just becomes stale data to be cleaned up on the next delete instead of wreaking havoc on subsequent concurrent snapshots.

original-brownbear committed:
make it a fragment

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class ShardGenerations implements ToXContent {
```
Fair point :) made it a fragment.

original-brownbear committed:
renamings

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -284,6 +316,8 @@ public IndexId resolveIndexId(final String indexName) {
         return snapshotIndices;
     }
 
+    public static final String SHARDS = "shards";
```
sure :)

original-brownbear commented:
Thanks Yannick, I htink I addressed all points form the first round now :)

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -426,14 +468,38 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                         assert indexId != null;
                         indexSnapshots.put(indexId, snapshotIds);
                     }
+                } else if (SHARD_GENERATIONS.equals(field)) {
```
is it possible to put this parsing code into `ShardGenerations`?

original-brownbear replied:
I really wanted to have the lookup by `IndexId`, specifically do avoid weirdness like the index name thing we have above. I can't really do that without the context of the index -> indexId mapping and since we don't have guarantees on the parsing order I also couldn't pass that mapping to a method in `ShardGenerations` to have it available during deserialization.

tlrx replied:
Is there a reason why shards generations are not contained within `indices` at the same level of `snapshots`? We could avoid to double check the index id/snapshot/shard gens coherency

original-brownbear replied:
@tlrx No there isn't :) I'll flatten the gens logic into that field then. That should be much nicer!

original-brownbear replied:
Moved to this form of serialization now :) Much nicer imo, let me know if this is what you had in mind.

original-brownbear committed:
fix and test handling same index name for different indices

original-brownbear committed:
fix logging

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

tlrx commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -78,19 +82,30 @@
         this.indices = Collections.unmodifiableMap(indexSnapshots.keySet().stream()
             .collect(Collectors.toMap(IndexId::getName, Function.identity())));
         this.indexSnapshots = Collections.unmodifiableMap(indexSnapshots);
+        this.shardGenerations = shardGenerations;
```
Add Objects.requireNonNull()?

tlrx commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -205,7 +231,10 @@ public RepositoryData removeSnapshot(final SnapshotId snapshotId) {
 
-        return new RepositoryData(genId, newSnapshotIds, newSnapshotStates, indexSnapshots);
+        return new RepositoryData(genId, newSnapshotIds, newSnapshotStates, indexSnapshots,
+            ShardGenerations.builder().addAll(shardGenerations).addAll(updatedShardGenerations)
```
Maybe we could build thing alongside with `indexSnapshots` in the `for` loop and avoid the extra `retainIndices` method on the builder that is only used here?

original-brownbear replied:
Then we have to somehow expose iteration over the generations object. I tried hiding the logic for that a little here. I know it's slightly more code but it seems a little easier to follow?

original-brownbear committed:
flatten delete method into one

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,189 @@
+         private final Map<IndexId, Map<Integer, String>> raw = new HashMap<>();
+
+         public Builder add(IndexId indexId, int shardId, String generation) {
+             raw.computeIfAbsent(indexId, i -> new HashMap<>()).put(shardId, generation);
```
Currently, the ability to override is actually used when merging two sets of generations.
I have to admit, I don't like the current strategy for all these computations all that much, but I didn't really find a nicer way of serializing things either and didn't want to just work from raw maps.

original-brownbear committed:
some Cr comments

original-brownbear committed:
adjust serialization

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:
```diff
@@ -412,14 +490,44 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+                            final String indexId = parser.currentName();
+                            if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
```
Alrighty, did use that now :)

original-brownbear commented:
Thanks @ywelsch and @tlrx . I think I've addressed all your comments now :)

ywelsch commented on server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -3761,6 +3771,65 @@ public void testRestoreIncreasesPrimaryTerms() {
         }
     }
 
+    public void testSnapshotDifferentIndicesBySameName() {
```
If we reduce the number of primaries, can this lead to orphaned shard folders when deleting the last snapshot that was referencing the prior index with the higher count of primaries?

original-brownbear replied:
You're right ... this needs some special case handling now. Otherwise we write an empty index-${uuid} blob that isn't referencing anything now. Will fix that tomorrow.

This whole index-name business is getting pretty weird ... but I don't see a clean way of changing the lookup to be by index-uuid either so :(

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -688,6 +748,20 @@ public void finalizeSnapshot(final SnapshotId snapshotId,
+        final int prefixPathLen = basePath().buildAsString().length();
+        updatedRepositoryData.shardGenerations().obsoleteShardGenerations(existingRepositoryData.shardGenerations()).forEach(
+            (indexId, gens) -> gens.forEach((shardId, oldGen) -> toDelete.add(
+                shardContainer(indexId, shardId).path().buildAsString().substring(prefixPathLen) + INDEX_FILE_PREFIX + oldGen)));
```
Sure handle that in a follow up (wanted to do that anyway since exposing the real path is a bit risky).

We do have this tested properly  I think. It passes all 3rd party tests (which all include cleanup tests that make sure deletes actually worked out).

ywelsch replied:
Should we extend the repo consistency check so that all the shard generations mentioned in the repository data exist?

original-brownbear replied:
Yea def. we should. I'll add that.

original-brownbear replied:
Ok I addressed this now in 1dffe60730d together with addressing the issue of reducing shard count by introducing special shard generations for deleted and known-to-not-exist shards.
Hope the added documentation makes things clear.

Also, I found a "bug" in the existing implementation that unfortunately I didn't find a clean way to solve here either. See the TODO in that commit, we currently fail to delete empty shard directories (just the directory itself not the contents). I didn't fix that here since it's a very minor issue, only affects FS+HDFS repos and is very tricky to fix cleanly.

original-brownbear committed:
Some CR comments addressed

original-brownbear committed:
Some CR comments addressed

original-brownbear committed:
add comment

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,167 @@
+     * @param previous Previous {@code ShardGenerations}
+     * @return Map of obsolete shard index generations in indices that are still tracked by this instance
+     */
+    public Map<IndexId, Map<Integer, String>> obsoleteShardGenerations(ShardGenerations previous) {
```
Done :)

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1224,50 +1312,41 @@ public String toString() {
-        BlobStoreIndexShardSnapshots snapshots = tuple.v1();
-        long fileListGeneration = tuple.v2();
-
+    private ShardSnapshotMetaDeleteResult deleteShardSnapshot(Set<SnapshotId> survivingSnapshots, IndexId indexId,
```
Done :)

original-brownbear committed:
refactor for easier readability

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -384,47 +388,154 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
      */
     private void doDeleteShardSnapshots(SnapshotId snapshotId, long repositoryStateId, Map<String, BlobContainer> foundIndices,
-                                        Map<String, BlobMetaData> rootBlobs, RepositoryData repositoryData,
+                                        Map<String, BlobMetaData> rootBlobs, RepositoryData repositoryData, Version version,
```
I know its 2 against 1 but I'm not sure I fully agree here tbh.

1. This has some issues:

```
                try {
                    blobContainer().deleteBlobsIgnoringIfNotExists(resolveFilesToDelete(snapshotId, deleteResults));
                    afterCleanupsListener.onResponse(null);
                } catch (Exception e) {
                    logger.warn(() -> new ParameterizedMessage("[{}] Failed to delete some blobs during snapshot delete", snapshotId), e);
                }
```

This won't work. We need to ensure that the listener is called even if the delete throws an exception. We'd haveto move the listener resolving into a `finally`.

2. This introduces a lot of duplication in precisely the kind of spot like the above one, where the error handling and making sure each listener gets called really matter. 
Also, the more we duplicate in these spots the less the test coverage we have that will catch the mistake above in the non-BwC path but won't catch it in the BwC path will help us out (this was the one of the motivations behind keeping things this dry, even if it's a little messier to read). The BwC path above has zero test coverage now for error conditions and it's gonna be pretty involved to code up BwC tests for it. 

=> Could I maybe warm your heart for 90c1df29f65 ? That factors out the shard meta update steps and computing the unreferenced blobs, gets rid of one listener but retains all the exception handling guarantees via the `ActionListener` framework and gives us proper test coverage?

tlrx replied:
> Could I maybe warm your heart for [90c1df2](https://github.com/elastic/elasticsearch/commit/90c1df29f65329dc41a21293c0c15564bfceda5e) ?

Thanks Armin, this is better but I still think we should improve this. I've LGTM the PR but I don't think it's very maintainable and keeping the dual logic into listeners is frankly confusing. Also, it adds complexity (I think) for the moment we'll have to remove the bwc logic all together.

ywelsch replied:
I think the duplication is ok, as it's just there, nicely isolated for the 7.x BWC path. In master, this code can go away. I don't expect much changes in this high-level logic (it just coordinates what is happening after what, the actual logic is implemented in helper methods). The important thing is that the series of actions is well understood. With the more callback-oriented approach that you've chosen, it's very difficult for me to understand the order of the various steps that the deletion is doing.

original-brownbear committed:
Add new shard placeholder

original-brownbear committed:
still need that bwc

original-brownbear committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
revert needless change

tlrx commented on server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -1639,7 +1641,15 @@ public void testSnapshotWithMissingShardLevelIndexFile() throws Exception {
+                            .filter(file -> file.getFileName().toString().startsWith(BlobStoreRepository.INDEX_FILE_PREFIX))
+                            .findFirst()
+                            .orElseThrow(() -> new AssertionError("Failed to find shard index blob"))
+                            .getFileName().toString().substring(BlobStoreRepository.INDEX_FILE_PREFIX.length());
```
Maybe return full absolute path (or relative file name) instead  and pass it to `Files.delete()`?

tlrx commented on test/framework/src/main/java/org/elasticsearch/index/shard/RestoreOnlyRepository.java:
```diff
@@ -89,20 +90,22 @@ public IndexMetaData getSnapshotIndexMetaData(SnapshotId snapshotId, IndexId ind
-        map.put(new IndexId(indexName, "blah"), emptySet());
-        return new RepositoryData(EMPTY_REPO_GEN, Collections.emptyMap(), Collections.emptyMap(), map);
+        final IndexId indexId = new IndexId(indexName, "blah");
+        map.put(indexId, emptySet());
```
This is never used?

tlrx commented on x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java:
```diff
@@ -78,21 +81,22 @@
-                                 List<SnapshotShardFailure> shardFailures, long repositoryStateId, boolean includeGlobalState,
-                                 MetaData metaData, Map<String, Object> userMetadata, ActionListener<SnapshotInfo> listener) {
+    public void finalizeSnapshot(SnapshotId snapshotId, ShardGenerations shardGenerations, long startTime, String failure,
+                                         int totalShards, List<SnapshotShardFailure> shardFailures, long repositoryStateId,
```
nit: indentation is wrong

tlrx reviewed (commented):
Thanks Armin, I can feel the pain of this PR by looking at the code :sweat_smile: 

I left more minor comments, I'm still digesting the changes in BlobStoreRepository.

original-brownbear committed:
track snapshot format in SNapshotsInProgress.Entry

original-brownbear committed:
fix indent

original-brownbear committed:
remove dead code

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -384,47 +388,154 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            executor.execute(ActionRunnable.wrap(deleteIndexMetaDataListener, deleteIdxMetaListener -> {
+                final IndexMetaData indexMetaData;
+                try {
+                    indexMetaData = getSnapshotIndexMetaData(snapshotId, indexId);
```
@tlrx that was my optimistic plan but it turns out this is impossible thanks to the fact that we associated index name (and not uuid) with the index id in the repository.
-> if we have 10 generations for an `IndexId` then we could have multiple indices by the same name snapshotted that have different shard counts but only on of those indices actually has a guarantee of having `10` shards :(
I think we can address this in an upcoming change though when dealing with the incrementality optimization in https://github.com/elastic/elasticsearch/issues/45736 where we will introduce some stricter accounting of what shard paths belong to what snapshots anyway.

tlrx replied:
Yes, it took me some time to reach the same conclusion as you, I've been confused by the IndexId/Index name resolution in repository data.

original-brownbear committed:
fix indent

original-brownbear committed:
unmodifiable map

original-brownbear committed:
nit empty map

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,167 @@
+
+public final class ShardGenerations {
+
+    public static final ShardGenerations EMPTY = ShardGenerations.builder().build();
```
Sure, I went with `Collections.emptyMap()` though -> less backporting pain :)

original-brownbear committed:
be less stupid when deleting

tlrx commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -384,47 +388,159 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            //
+            // Note: If we fail updating any of the individual shard paths, none of them are changed since the newly created
+            //       index-${gen_uuid} will not be referenced by the existing RepositoryData and new RepositoryData is only
+            //       written if all shard paths have been successfully updated.;
```
nit: extra `;`

tlrx commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -384,47 +388,159 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+        } else {
+            // Writing the new repository data first, without tracking any shard generations in the BwC path
+            writeNewRepoData.accept(ShardGenerations.EMPTY);
+            // We've already written the new RepositoryData so updating all shard-level medata will mean that all metadata is updated
```
nit: typo `medata`

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:
```diff
@@ -91,12 +91,13 @@ public String toString() {
         private final ImmutableOpenMap<String, List<ShardId>> waitingIndices;
         private final long startTime;
         private final long repositoryStateId;
+        private final boolean newFormat;
```
This might not age well without a comment and/or an explicit name. Maybe something like `useShardGenerations`?

original-brownbear replied:
jup :) that sounds better

tlrx reviewed (approved):
LGTM.

Thanks a lot Armin for this painful change. I've look deeper how shard gens play with snapshot deletion and creation and with the latest changes (thanks to Yannick's review, but I spotted some of them too) I don't see any pitfall in this change.

I still think that the `doDeleteShardSnapshots()` method should be improved as it is not maintainable in its current state, but I don't want to hold this PR for longer because of this. We can improve this later, and looking at the 1500 LoC of BlobStoreRepository class I think it would benefit to be split to distinguish between root/high level and low/shard level logic.

original-brownbear committed:
CR: correctly deal with lowering shard count on constant index name

original-brownbear committed:
address nits

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc (non snapshot related ML issue :))

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
add generation to logging

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -962,17 +1042,22 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
         try {
             logger.debug("[{}] [{}] snapshot to [{}] ...", shardId, snapshotId, metadata.name());
 
+            final String generation = snapshotStatus.generation();
```
Done :)

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample

original-brownbear commented:
@ywelsch I added stronger consistency checks and functionality to the generations that allow differentiating between "definitely new shard" and "maybe BwC shard", added an IT for (and fixed) the cleanup of shards in the multiple-indices-of-the-same-name scenario.
Also, added a old-style flag to the `SnapshotsInProgress.Entry` => I think I've addressed all your comments and this is ready for another round.

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -862,7 +941,7 @@ protected void writeIndexGen(final RepositoryData repositoryData, final long exp
         final String indexBlob = INDEX_FILE_PREFIX + Long.toString(newGen);
         logger.debug("Repository [{}] writing new index generational blob [{}]", metadata.name(), indexBlob);
-        writeAtomic(indexBlob, BytesReference.bytes(repositoryData.snapshotsToXContent(XContentFactory.jsonBuilder())), true);
+        writeAtomic(indexBlob, BytesReference.bytes(repositoryData.snapshotsToXContent(XContentFactory.jsonBuilder(), version)), true);
```
Can we assert that repositoryData.shardGenerations does not contain `NEW_SHARD_GEN` entries? `NEW_SHARD_GEN` should not be written to disk either?

original-brownbear replied:
I would in fact write `NEW_SHARD_GEN` to disk. You'd want this in case a shard snapshot was started on a data node but the node left during snapshotting. 
If you write this information to disk, a subsequent snapshot operation knows to not try and list out the shard directory (as `null` would imply if we were to just prune these values before writing).
It's really hard to do this in subsequent snapshots in `SnapshotsService` when creating the snapshot otherwise since we don't track the shard count for each index (and can't easily do that either since `RepositoryData` works with index names for now).

ywelsch replied:
> It's really hard to do this in subsequent snapshots in SnapshotsService when creating the snapshot otherwise since we don't track the shard count for each index (and can't easily do that either since RepositoryData works with index names for now).

ok, thanks for the explanation. That's really unfortunate. Can we assert then that we're never writing out a shard state metadata file with a `_new` prefix?

original-brownbear replied:
Added assertion :)

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1296,6 +1382,29 @@ private BlobStoreIndexShardSnapshot loadShardSnapshot(BlobContainer shardContain
+                                                                                          BlobContainer shardContainer,
+                                                                                          @Nullable String generation) throws IOException {
+        if (generation != null) {
+            if (generation.equals(ShardGenerations.NEW_SHARD_GEN)) {
```
I would prefer  to move this logic to the caller (there is just one where this is applicable) and assert in this method that the generation is not one of these special generation types

original-brownbear replied:
See the previous comments on `NEW_SHARD_GEN` persistency -> There's two where it's applicable. The delete can also see this when there's only a failed snapshot (e.g. where the node left and never updated master) for the shard.

ywelsch replied:
ok

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -229,6 +230,7 @@ private void startNewSnapshots(SnapshotsInProgress snapshotsInProgress) {
         final String localNodeId = clusterService.localNode().getId();
         for (SnapshotsInProgress.Entry entry : snapshotsInProgress.entries()) {
             final State entryState = entry.state();
+            final Version version = SnapshotsService.compatibleVersion(entry);
```
let's use and propagate the boolean here, not this version.

original-brownbear replied:
Done

ywelsch commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java:
```diff
@@ -118,6 +123,27 @@ private static void assertIndexGenerations(BlobContainer repoRoot, long latestGe
+                for (int i = 0; i < gens.size(); i++) {
+                    final String generation = gens.get(i);
+                    assertThat(generation, not(ShardGenerations.DELETED_SHARD_GEN));
+                    if (generation != null && generation.equals(ShardGenerations.NEW_SHARD_GEN) == false) {
```
assertConsistency is for what's on disk? The repositoryData should not contain `NEW_SHARD_GEN` entries?

original-brownbear replied:
> assertConsistency is for what's on disk?

Yes

> The repositoryData should not contain NEW_SHARD_GEN entries?

See above, I think it should contain `NEW_SHARD_GEN` entries.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java:
```diff
@@ -118,6 +123,27 @@ private static void assertIndexGenerations(BlobContainer repoRoot, long latestGe
+            final List<String> gens = shardGenerations.getGens(index);
+            if (gens.isEmpty() == false) {
+                final BlobContainer indexContainer = indicesContainer.children().get(index.getId());
+                final Map<String, BlobContainer> shardContainers = indexContainer.children();
```
should we also check which files should not be there?

original-brownbear replied:
We never actually write any blobs to the root of the `indices` folder that's why there's no check for that. But sure I can add one (in a separate PR maybe since it's totally unrelated to this effort?) to make sure it stays that way :)

original-brownbear replied:
Ah ok. See the TODO `assertConsistency`, I was gonna add some more inverse checks eventually for sure. This is mainly dependent on adding shard cleanups to the cleanup endpoint (we run it before the consistency check to make sure the indices folders match up in tests that do failure scenarios that leak files but don't corrupt the repo).
If we want to do these checks on the shard level we need to beef up the cleanup endpoint first or we'd have to exclude part of the check for failure scenarios. That's the only reason thats's not in here yet :)

ywelsch replied:
ok, makes sense

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java:
```diff
@@ -0,0 +1,213 @@
+            for (int i = 0; i < Math.min(oldGens.size(), updatedGenerations.size()); i++) {
+                final String oldGeneration = oldGens.get(i);
+                final String updatedGeneration = updatedGenerations.get(i);
+                if (updatedGeneration != null && oldGeneration != null && oldGeneration.equals(updatedGeneration) == false) {
```
can you add a comment here, saying why we only override info for cases where we have oldGeneration as well as updatedGeneration?

original-brownbear replied:
Done

ywelsch reviewed (changes requested):
I've left some more comments, we're converging though.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
add comment

original-brownbear committed:
flag not version

original-brownbear committed:
flatter ordered blobstore delete

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -1070,27 +1155,35 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
                         + "] when deleting index-N blobs " + blobsToDelete;
+                }
+                try {
+                    final BlobStoreIndexShardSnapshots updatedSnapshots = new BlobStoreIndexShardSnapshots(newSnapshotsList);
```
We actually have that now so I could add an assertion about the generation :)

original-brownbear commented:
@ywelsch how about 77eeb7d9666  for making the order clearer? :) No we really only have callbacks that force you to look back a few lines to get the execution order in a single place and I didn't ahve to mix `try-catch` with listener handling.

ywelsch commented:
> how about 77eeb7d for making the order clearer?

I find it still as complex to understand. It still defines 4 abstract things and plugs them together in various ways. It's impossible to tell which listener will be invoked, and how often it is invoked.

Here's another patch which imho makes it easier to understand what's going on:

```
diff --git a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
index c5bd70f586b..7e9234436bb 100644
--- a/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
+++ b/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java
@@ -62,7 +62,6 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.common.util.concurrent.AbstractRunnable;
-import org.elasticsearch.common.util.concurrent.ListenableFuture;
 import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
 import org.elasticsearch.common.xcontent.XContentFactory;
@@ -109,7 +108,6 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Executor;
-import java.util.concurrent.ExecutorService;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -390,78 +388,77 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
      */
     private void doDeleteShardSnapshots(SnapshotId snapshotId, long repositoryStateId, Map<String, BlobContainer> foundIndices,
                                         Map<String, BlobMetaData> rootBlobs, RepositoryData repositoryData, boolean writeShardGens,
-                                        ActionListener<Void> listener) {
-
-        final ExecutorService executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
-
-        // Listener to invoke once an updated ShardGenerations instance is ready to be written
-        final ListenableFuture<ShardGenerations> shardGenerationsListener = new ListenableFuture<>();
-
-        // Listener to invoke once we've written the latest repository data and updated all shard metadata.
-        // Any failure past this point will only be logged.
-        final ListenableFuture<Collection<ShardSnapshotMetaDeleteResult>> afterUpdateAllMetadata = new ListenableFuture<>();
-
-        final ListenableFuture<RepositoryData> repositoryDataWrittenListener = new ListenableFuture<>();
+                                        ActionListener<Void> listener) throws IOException {
 
         if (writeShardGens) {
-            // New path that updates the pointer to each deleted shard's generation in the root RepositoryData
-
-            // Once we are done removing the snapshot from the shard-level metadata of each affected shard we can update the repository
-            // metadata as follows:
+            // First write the new shard state metadata (with the removed snapshot) and compute deletion targets
+            final StepListener<Collection<ShardSnapshotMetaDeleteResult>> writeShardMetaDataAndComputeDeletesStep = new StepListener<>();
+            writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, true, writeShardMetaDataAndComputeDeletesStep);
+            // Once we have put the new shard-level metadata into place, we can update the repository metadata as follows:
             // 1. Remove the snapshot from the list of existing snapshots
             // 2. Update the index shard generations of all updated shard folders
             //
             // Note: If we fail updating any of the individual shard paths, none of them are changed since the newly created
             //       index-${gen_uuid} will not be referenced by the existing RepositoryData and new RepositoryData is only
             //       written if all shard paths have been successfully updated.
-            final StepListener<Collection<ShardSnapshotMetaDeleteResult>> afterUpdateShardMeta = new StepListener<>();
-
-            writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, true, afterUpdateShardMeta);
-
-            afterUpdateShardMeta.whenComplete(res -> {
-                final ShardGenerations.Builder builder = ShardGenerations.builder();
-                for (ShardSnapshotMetaDeleteResult newGen : res) {
-                    builder.put(newGen.indexId, newGen.shardId, newGen.newGeneration);
-                }
-                shardGenerationsListener.onResponse(builder.build());
-                repositoryDataWrittenListener.addListener(ActionListener.delegateFailure(listener,
-                    (v, l) -> afterUpdateAllMetadata.onResponse(res)), executor);
+            final StepListener<RepositoryData> writeUpdatedRepoDataStep = new StepListener<>();
+            writeShardMetaDataAndComputeDeletesStep.whenComplete(deleteResults -> {
+                    final ShardGenerations.Builder builder = ShardGenerations.builder();
+                    for (ShardSnapshotMetaDeleteResult newGen : deleteResults) {
+                        builder.put(newGen.indexId, newGen.shardId, newGen.newGeneration);
+                    }
+                    final RepositoryData updatedRepoData = repositoryData.removeSnapshot(snapshotId, builder.build());
+                    writeIndexGen(updatedRepoData, repositoryStateId, true);
+                    writeUpdatedRepoDataStep.onResponse(updatedRepoData);
+                }, writeUpdatedRepoDataStep::onFailure);
+            // Once we have updated the repository, run the clean-ups
+            writeUpdatedRepoDataStep.whenComplete(updatedRepoData -> {
+                // Run unreferenced blobs cleanup in parallel to shard-level snapshot deletion
+                final ActionListener<Void> afterCleanupsListener =
+                    new GroupedActionListener<>(ActionListener.wrap(() -> listener.onResponse(null)), 2);
+                asyncCleanupUnlinkedRootAndIndicesBlobs(foundIndices, rootBlobs, updatedRepoData, afterCleanupsListener);
+                asyncCleanupUnlinkedShardLevelBlobs(snapshotId, writeShardMetaDataAndComputeDeletesStep.result(), afterCleanupsListener);
             }, listener::onFailure);
         } else {
-            // Writing the new repository data first, without tracking any shard generations in the BwC path
-            shardGenerationsListener.onResponse(ShardGenerations.EMPTY);
-            // We've already written the new RepositoryData so updating all shard-level metadata will mean that all metadata is updated
-            repositoryDataWrittenListener.addListener(ActionListener.delegateFailure(listener,
-                (v, l) -> writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, false, afterUpdateAllMetadata)),
-                executor);
+            // Write the new repository data first (with the removed snapshot), using no shard generations
+            final RepositoryData updatedRepoData = repositoryData.removeSnapshot(snapshotId, ShardGenerations.EMPTY);
+            writeIndexGen(updatedRepoData, repositoryStateId, false);
+            // Run unreferenced blobs cleanup in parallel to shard-level snapshot deletion
+            final ActionListener<Void> afterCleanupsListener =
+                new GroupedActionListener<>(ActionListener.wrap(() -> listener.onResponse(null)), 2);
+            asyncCleanupUnlinkedRootAndIndicesBlobs(foundIndices, rootBlobs, updatedRepoData, afterCleanupsListener);
+            writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, false, ActionListener.map(afterCleanupsListener,
+                deleteResults -> {
+                    asyncCleanupUnlinkedShardLevelBlobs(snapshotId, deleteResults, afterCleanupsListener);
+                    return null;
+                }));
         }
+    }
+
+    private void asyncCleanupUnlinkedRootAndIndicesBlobs(Map<String, BlobContainer> foundIndices, Map<String, BlobMetaData> rootBlobs,
+        RepositoryData updatedRepoData, ActionListener<Void> listener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(
+            listener,
+            l -> cleanupStaleBlobs(foundIndices, rootBlobs, updatedRepoData, ActionListener.map(l, ignored -> null))));
+    }
 
-        shardGenerationsListener.addListener(ActionListener.wrap(shardGens -> {
-            final RepositoryData updatedRepoData = repositoryData.removeSnapshot(snapshotId, shardGens);
-            // Write out new RepositoryData
-            writeIndexGen(updatedRepoData, repositoryStateId, writeShardGens);
-            repositoryDataWrittenListener.onResponse(updatedRepoData);
-        }, listener::onFailure), executor);
-
-        // Listener that resolves the given listener passed to this method once both cleanup steps have completed.
-        // Any failures encountered by this listener are ignored as they were already logged by the throwing code.
-        final ActionListener<Void> afterCleanupsListener =
-            new GroupedActionListener<>(ActionListener.wrap(() -> listener.onResponse(null)), 2);
-
-        // Run unreferenced blobs cleanup if we were able to write updated repository data
-        repositoryDataWrittenListener.addListener(
-            ActionListener.wrap(updatedRepoData -> cleanupStaleBlobs(
-                foundIndices, rootBlobs, updatedRepoData, ActionListener.map(afterCleanupsListener, ignored -> null)), e -> {}), executor);
-
-        afterUpdateAllMetadata.addListener(ActionListener.runAfter(
-            ActionListener.wrap(
-                // Now that all metadata (RepositoryData at the repo root as well as index-N blobs in all shard paths)
-                // has been updated we can execute the delete operations for all blobs that have become unreferenced as a result
-                deleteResults -> blobContainer().deleteBlobsIgnoringIfNotExists(resolveFilesToDelete(snapshotId, deleteResults)),
-                // Any exceptions after we have updated the root level RepositoryData are only logged but won't fail the delete request
-                e -> logger.warn(
-                    () -> new ParameterizedMessage("[{}] Failed to delete some blobs during snapshot delete", snapshotId), e)),
-            () -> afterCleanupsListener.onResponse(null)), executor);
+    private void asyncCleanupUnlinkedShardLevelBlobs(SnapshotId snapshotId, Collection<ShardSnapshotMetaDeleteResult> deleteResults,
+        ActionListener<Void> listener) {
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(
+            listener,
+            l -> {
+                try {
+                    blobContainer().deleteBlobsIgnoringIfNotExists(resolveFilesToDelete(
+                        snapshotId,
+                        deleteResults));
+                    listener.onResponse(null);
+                } catch (Exception e) {
+                    logger.warn(
+                        () -> new ParameterizedMessage("[{}] Failed to delete some blobs during snapshot delete", snapshotId),
+                        e);
+                    throw e;
+                }
+            }));
     }
 
     // updates the shard state metadata for shards of a snapshot that is to be deleted. Also computes the files to be cleaned up.
```

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
CR: Yannicks patch

original-brownbear committed:
slight change

original-brownbear commented:
Thanks @ywelsch, I like that patch :) Applied in 1bdbf2c with minor change https://github.com/elastic/elasticsearch/pull/46250/commits/68e407cbff603d9ffcc78a495bb968ad6ac4fdbc (to make it a little more obvious what fails the overall request and what doesn't) now :)

original-brownbear commented:
Jenkins run elasticsearch-ci/2  (unrelated xpack security failure)

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -388,47 +390,171 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
+            final ActionListener<Void> afterCleanupsListener =
+                new GroupedActionListener<>(ActionListener.wrap(() -> listener.onResponse(null)), 2);
+            asyncCleanupUnlinkedRootAndIndicesBlobs(foundIndices, rootBlobs, updatedRepoData, afterCleanupsListener);
+            writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, false, ActionListener.map(afterCleanupsListener,
```
The listener here returns as soon as `asyncCleanupUnlinkedShardLevelBlobs` returns, which is not what we want. Instead we could do:

```
writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, false,
                new ActionListener<>() {
                    @Override public void onResponse(Collection<ShardSnapshotMetaDeleteResult> deleteResults) {
                        asyncCleanupUnlinkedShardLevelBlobs(snapshotId, deleteResults, afterCleanupsListener);
                    }

                    @Override public void onFailure(Exception e) {
                        afterCleanupsListener.onFailure(e);
                    }
                });
```

or alternatively:

```
final StepListener<Collection<ShardSnapshotMetaDeleteResult>> writeMetaAndComputeDeletesStep = new StepListener<>();
writeUpdatedShardMetaDataAndComputeDeletes(snapshotId, repositoryData, false, writeMetaAndComputeDeletesStep);
writeMetaAndComputeDeletesStep.whenComplete(deleteResults ->
    asyncCleanupUnlinkedShardLevelBlobs(snapshotId, deleteResults, afterCleanupsListener), afterCleanupsListener::onFailure);
```

ywelsch reviewed (approved):
I've left one important comment that needs addressing, o.w. LGTM

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear committed:
CR: fix listener fallthrough

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-incrementality-snapshots

original-brownbear commented:
Thanks so much for grinding through this @tlrx and @ywelsch, much appreciated !
